Python/Module6_Testing/Hypothesis.md

```diff
@@ -0,0 +1,979 @@
+
+import hypothesis.strategies as st
+from hypothesis import given, note
+
+# a list of all printable non-vowel characters
+_not_vowels = "".join([l for l in printable if l.lower() not in set("aeiouy")])
+
+
+@given(
+    not_vowels=st.text(alphabet=_not_vowels),
+    vowels_but_not_ys=st.text(alphabet="aeiouAEIOU"),
+    ys=st.text(alphabet="yY"),
+)
+def test_count_vowels_hypothesis(not_vowels, vowels_but_not_ys, ys):
+    """
+    Constructs an input string with a known number of:
+       - non-vowel characters
+       - non-y vowel characters
+       - y characters
+    
+    and thus, by constructions, we can test that the output
+    of `count_vowels` agrees with the known number of vowels
+    """
+    # list of characters
+    letters = list(not_vowels) + list(vowels_but_not_ys) + list(ys)
+    
+    # We need to shuffle the ordering of our characters so that
+    # our input string isn't unnaturally patterned; e.g. always 
+    # have its vowels at the end
+    shuffle(letters)
+    in_string = "".join(letters)
+    
+    # Hypothesis provides a `note` function that will print out
+    # whatever input you give it, but only in the case that the
+    # test fails.
+    # This way we can see the exact string that we fed to `count_vowels`,
+    # if it caused our test to fail
+    note("in_string: " + in_string)
+    
+    # testing that `count_vowels` produces the expected output
+    # both including and excluding y's in the count
+    assert count_vowels(in_string, include_y=False) == len(vowels_but_not_ys)
+    assert count_vowels(in_string, include_y=True) == len(vowels_but_not_ys) + len(ys)
+```
+<!-- #endregion -->
+
+
+**Using the `.map` method to create a sorted list: Solution**
+
+Using the `.map()` method, construct a Hypothesis strategy that produces a sorted list of integers.
+Generate some examples from your strategy and check that they are sorted (we may have to generate quite a few examples to see a diverse set of values)
+
+<!-- #region -->
+```python
+# Note that the built-in `sorted` function can be supplied 
+# directly to the `.map()` method - there is no need to define
+# a function or use a lambda expression here
+sorted_list_of_ints = st.lists(st.integers()).map(sorted)
+```
+```python
+>>> sorted_list_of_ints.example()
+[-27120, 97, 12805]
+```
+<!-- #endregion -->
+
+
+**Getting creative with the `.map` method: Solution**
+
+Construct a Hypothesis strategy that produces either the string `"cat"` or the string `"dog"`.
+Then, write a test that uses this strategy;
+it should simply test that either `"cat"` or `"dog"` was indeed produced by the strategy.
+Run the test
+
+<!-- #region -->
+```python
+# We "hijack" the `st.booleans()` strategy, which only generates 
+# `True` or `False`, and use the `.map` method to transform these
+# two outputs to `"cat"` or `"dog"`.
+#
+# This is only one of many ways that you could have created this
+# strategy
+cat_or_dog = st.booleans().map(lambda x: "cat" if x else "dog")
```

Comment:
I'd prefer to show `st.sampled_from(["cat", "dog"])` as the primary example, with the remapped booleans as a curiosity.